sql/function/normalize/normalizer-option.sql

```diff
@@ -0,0 +1 @@
+SELECT pgroonga_normalize('aBcDe 123', 'NormalizerAuto');
```

Comment:
`NormalizerAuto` is meaningless because `NormalizerAuto` is the default.
How about using https://github.com/groonga/groonga-normalizer-mysql ?

Reply:
> How about using https://github.com/groonga/groonga-normalizer-mysql ?

It will break the test suite on many systems, since `groonga-normalizer-mysql` is not installed by default.

AFAICT, currently there is no such mechanism to skip tests based on system conditions. Yes, in theory, we can accomplish this goal by introducing something like this to `pgroonga.mk`:

```make
ifndef HAVE_MYSQL_NORMALIZER
  REGRESS = $(filter-out function/normalize/normalizer-option, $(REGRESS))'
notdef
```

or even:

```make
CMDOUT = $(shell psql -A -t -c "SELECT pgroonga_command('normalizer_list')")
ifeq (,$(findstring NormalizerAuto,$(CMDOUT)))
  REGRESS = $(filter-out function/normalize/normalizer-option, $(REGRESS))'
endif
```

But I'm so sure if this look like a terribly good solution.

What do you think about this? Or do you know a better approach?

Reply:
最初のdescriptionが日本語なので、ここのコメントも日本語で大丈夫ですよね……？

自分はMroongaにもPGroongaにもPostgreSQLにも明るくないのですが、もし未定義の引数が渡される（例えば引数が1個までの関数に2個目の引数を渡すとか）とエラーになるという仕様なのであれば、このテストでも「第2引数を受け付けてエラーにならないかどうか」だけはテストできていると言えると思います。
また、未知のnormalizerを第2引数で指定されたらエラーになる仕様なのであれば、ここにさらに絶対にあり得ない変な名前のnormalizerを指定するテストを追加する事で、エラー検出機構のテストはできると思います。
となると、後の問題は、「指定された名前のnormalizerが使われるかどうかのテストをどうやってやるか」ということになります。
理想を言うと、通常のこれらのテストとは別のレイヤーで「`groonga-normalizer-mysql`がインストールされた環境、を用意する所までをsetupで行うようなテスト」を用意し、「このケースはそちらを参照せよ」のようにコメントを書いておくのがいいのではないか、と思うのですがどうでしょうか。
（この場合のデメリットは、このテスト1つのためだけに既存のテストの枠組みからはみ出す物を導入する羽目になる事と、テスト実行のコストが跳ね上がるという事でしょうか。）

Reply:
> また、未知のnormalizerを第2引数で指定されたらエラーになる仕様なのであれば、ここにさらに絶対にあり得ない変な名前のnormalizerを指定するテストを追加する事で、エラー検出機構のテストはできると思います。

こちらについては、すぐ上の [sql/function/normalize/invalid.sql](https://github.com/pgroonga/pgroonga/pull/56/files#diff-1878228e2bd60e7f11ba0e8c2d26db33) で対応済みです。

> 理想を言うと、通常のこれらのテストとは別のレイヤーで「groonga-normalizer-mysqlがインストールされた環境、を用意する所までをsetupで行うようなテスト」を用意し、「このケースはそちらを参照せよ」のようにコメントを書いておくのがいいのではないか、と思うのですがどうでしょうか。

これについては現状まったく仕組みがないので、このプルリク内で新しくテストの仕組みを導入するところまでやるのか？が一つの焦点だと思います。
また、PGXSの仕組みでこのテストを実現する方法があるんだろうか、という点で調査する必要があると思います。今のPGXSベースだとSQLを実行できるだけなので、任意の実行スクリプトをテスト内に埋め込む方法はないように思います。